lib/search-suggestions/index.jsx

```diff
@@ -0,0 +1,115 @@
+import React, { Component } from 'react';
+import PropTypes from 'prop-types';
+import { connect } from 'react-redux';
+import SmallSearchIcon from '../icons/search-small';
+import TagIcon from '../icons/tag';
+
+export class SearchSuggestions extends Component {
+  static displayName = 'SearchSuggestions';
+
+  static propTypes = {
+    onSearch: PropTypes.func.isRequired,
+    query: PropTypes.string.isRequired,
+    filteredTags: PropTypes.array.isRequired,
+    storeKeyHandler: PropTypes.func.isRequired,
+  };
+
+  state = {
+    selectedItem: 0,
+  };
+
+  componentDidMount() {
+    this.props.storeKeyHandler({
+      next: this.nextItem,
+      prev: this.prevItem,
+      select: this.selectItem,
+    });
+  }
+
+  componentDidUpdate() {
+    this.props.storeKeyHandler({
+      next: this.nextItem,
+      prev: this.prevItem,
+      select: this.selectItem,
+    });
+  }
+
+  nextItem = () => {
+    const { filteredTags } = this.props;
+    const { selectedItem } = this.state;
+    const newItem =
+      selectedItem === filteredTags.length
+        ? filteredTags.length
+        : selectedItem + 1;
+    this.setState({ selectedItem: newItem });
+  };
+
+  prevItem = () => {
+    const { selectedItem } = this.state;
+    const newItem = selectedItem === 0 ? 0 : selectedItem - 1;
+    this.setState({ selectedItem: newItem });
+  };
+
+  selectItem = () => {
+    const { selectedItem } = this.state;
+    const { query, onSearch, filteredTags } = this.props;
+    if (selectedItem === 0) {
+      onSearch(query);
+    } else {
+      onSearch(`tag:${filteredTags[selectedItem - 1].id}`);
+    }
+  };
+
+  render() {
+    const { query, onSearch, filteredTags } = this.props;
+    const { selectedItem } = this.state;
+    const screenReaderLabel = 'Search suggestions';
+    const shouldShowTagSuggestions = query.length > 2;
+
+    return (
+      <div className="search-suggestions" aria-label={screenReaderLabel}>
+        <div
+          id="query"
+          className={
+            selectedItem === 0
+              ? 'search-suggestion-row search-suggestion-row-selected'
+              : 'search-suggestion-row'
+          }
+          onClick={() => onSearch(query)}
+        >
+          <SmallSearchIcon />
+          <div className="search-suggestion">{query}</div>
+        </div>
+        {shouldShowTagSuggestions &&
+          filteredTags.map((tag, index) => (
+            <div
+              key={tag.id}
+              id={tag.id}
+              className={
+                selectedItem === index + 1
+                  ? 'search-suggestion-row search-suggestion-row-selected'
+                  : 'search-suggestion-row'
+              }
+              onClick={() => onSearch(`tag:${tag.id}`)}
```

Comment:
fun fact, not too important here, but can be useful especially since this event originates on a `div` - we can pass data through custom attributes and access them in the callback…

```js
<div
	key={tag.id}
	id={tag.id}
	data-tag-id={tag.id}
	onClick={this.doSearch}
>
```

```js
doSearch = ( { target: { dataset: { tagId } } } ) => this.props.onSearch( `tag:${ tagId }` ));
```

Reply:
Ooh this is something I have needed in attempting to translate my jQuery brain into React.